### PR TITLE
Add new args

### LINKED ISSUE #6: SHA256 (256-bit) checksums
Researching the ability to validate SHA256 (256-bit) checksums.

- SHA256 checksums are available from the servers. Just add `.sha256` to the end of the file's link.
- `sha256sum` seems to be a default app/program on most Linux distros.

Please provide input on what the script should be expected to do.

Should these checksum files be saved along side the ZIMs and kept? Should they be purged after validation? Should the script always validate checksums?

I'm just stuck on what the "flow" should be. Please help me visualize this.

### PR DESCRIPTION
**Pull request by jojo2357**
Can now specify min/max file size, and checksums, and skipping purge

Closes #6 #12 #11 #10

### CODE REVIEW

File: kiwix-zim.sh [hunk 1]
```diff
@@ -137,12 +142,16 @@ usage_example() {
...
+    echo '    -c, --calculate-checksum   Verifies that the downloaded files were not corrupted, but can take a while for large downloads.'
     echo '    -d, --disable-dry-run      Dry-Run Override.'
     echo '                               *** Caution ***'
     echo
     echo '    -h, --help                 Show this usage and exit.'
-    echo
     echo '    -p, --skip-purge           Skips purging any replaced ZIMs.'
+    echo '    -n <size>, --min-size      Minimum ZIM Size to be downloaded.'
+    echo '                               Specify units include M Mi G Gi, etc. See `man numfmt`'
+    echo '    -x <size> , --man-size     Maximum ZIM Size to be downloaded.'
```
Comment (Jaifroid) [inline kiwix-zim.sh:153]: @jojo2357 Should this be `--max-size` rather than `--man-size`?

**Comment by Jaifroid** [comment]
Great PR! On testing it in dry-run mode, setting the -x option doesn't seem to do anything (it still simulates downloading full English Wikipedia, for example). It would be good if the simulation reflected the actual archives to be downloaded. I guess this is because in dry-run mode the file size is not checked?

**Comment by Jaifroid** [comment]
Further info: I ran the command like this: `./kiwix-zim.sh /mnt/w/ --skip-purge -x 500M`. The skip-purge option worked (Purging Old ZIMs - `Skipped` in red), but in the list of simulated downloads I got:
```
  Download (mirror) : https://www.mirrorservice.org/sites/download.kiwix.org/zim/wikipedia/wikipedia_en_all_maxi_2023-04.zim
  ✓ Status : *** Simulated ***  ZIM doesn't exist on disk. Downloading...
```

**Comment by Jaifroid** [comment]
@jojo2357 Apologies, I was fooled by the fact that the script uses Git commands to reset itself to `main` before running... So ended up running `main` instead of the PR without realizing it.

**Comment by Jaifroid** [comment]
Had to comment out the block that switches back to `main`, and it works fine now:
```
  Download (mirror) : https://www.mirrorservice.org/sites/download.kiwix.org/zim/wikipedia/wikipedia_en_all_maxi_2023-04.zim
  ✓ Status : *** Simulated ***  ZIM larger than specified maximum size (maximum: 477Mi, download size: 94Gi). Skipping download.
```
Apologies for the confusion!

**Comment by Jaifroid** [comment]
OK, I think this is a legit issue with the PR: at least in the simulated run, it appears to want to purge archives that were not downloaded due to the max-size check (see screenshot). In practice the purge should fail because the new file had not been downloaded (there is a comment `# Before we actually purge, we want to check that the new ZIM downloaded and exists.`), but the simulation should reflect the fact that the archive won't be purged because not downloaded due to a user option.

image

File: kiwix-zim.sh (hunk 1 above)
Comment (jojo2357) [reply kiwix-zim.sh:153]: Yes 😂

**Comment by jojo2357** [comment]
Yea, the branch switching is a nuisance. If you clone my repo then it is easier to test, but you either have to commit *and* push to `origin` or just comment that block iff you make edits. It is easiest to just add a `false && ` to the first if.

As for the simulation lying, I believe i fixed it. In the event that the zim does *not* exist on disk, but is skipped due to size constraints, it will warn with extra detail in the simulation.

**Review by Jaifroid — APPROVED** [review]
Looks great to me. I tested the latest code in this PR and it works as described and as intended. Great work! 😊

My knowledge of bash scripting is a bit sketchy, so I didn't attempt to review in detail the code. I focused on testing it empirically (in simulated mode) against my relatively large ZIM library. I also did a real run with the --max-size and --skip-purge options.

**Comment by jojo2357** [comment]
Okay, I will update the readme tonight, so don't merge until I get a chance to revise it.

I'm glad to hear of the success, no checksums though? Definitely could use a test. Can your bittorrent intentionally corrupt a download?

**Comment by Jaifroid** [comment]
I didn't test the checksum option. Will have a go and report back.

**Comment by Jaifroid** [comment]
Just ran a test on this, and checksum function works fine!

Your comment about BitTorrent corrupting files leads me to wonder whether you intended the checksum function to test files that are not downloaded because they already exist on the disk? Currently of course it only tests files it has downloaded.

Do you intend that to be an option? It could be a useful option to check the integrity of a whole ZIM library against expected checksums as provided by the download server. However, it should probably be a separate issue/PR, and definitely a separate commandline option given the amount of time it could take.

**Comment by jojo2357** [comment]
Right now i only intend it as a way to verify the integrity of files dowloaded during the current run. Everything in this pr was easy to hook into the current code flow, and verifying all files doesnt fit nicely into the current flow, as i see it.

If I/we ever address #7, then some redesign will be necessary and would probably make verifying the entire library feasible. 

One consideration though: if you are constantly using this script to update, and using the checksum option, after a while you will know all your files were downloaded validly (because they were all updated once or more). What usecase then would be served to have this extra feature?

**Comment by Jaifroid** [comment]
I agree: definitely a separate issue and more complex to implement. NB I'm not advocating for this particularly: it was only prompted by your mention of BitTorrent.

The use case would be that the user may get downloads of larger files from other sources and applications. Particularly the case in low-bandwidth situations (we recently had an email from a user in Cuba where Kiwix archives are part of "el paquete", a download of offline data that Cubans swap on USB sticks). Such low-bandwidth situations are prevalent throughout much of the developing world outside of large metropolitan centres.

On the other hand, if a user is security-conscious, they probably know how to verify a file for themselves... And this tool is only likely to be used by enthusiasts and expert users.

**Comment by DocDrydenn** [comment]
Awesome work, guys. Thanks for the help. Just been swamped lately, so I apologize for my delays.

**Review by DocDrydenn — COMMENTED** [review]
Merged

**Comment by jojo2357** [comment]
> Okay, I will update the readme tonight, so don't merge until I get a chance to revise it.

Oh well, an extra commit will be needed. 

As for the other things: it might actually be easy to generate a `library.sha256` and then just run `sha256sum`. Doing it file-by-file may be tricky, but in the current flow it could actually be quite easy 🤔. I wouldnt delete them if it failed, just show the sum output. 

Maybe. Im not looking at it right now.

**Comment by Jaifroid** [comment]
> Awesome work, guys. Thanks for the help. Just been swamped lately, so I apologize for my delays.

Thanks for your great script! It would be good to find a way to promote it a bit more. Would you like me to post about it on the Kiwix Mastodon account and/or add a new post about the updates on the Kiwix Reddit (or you might prefer to do the latter)?

**Comment by DocDrydenn** [comment]
I posted it to a couple of subs on Reddit a while back, but got very little "buzz"... just figured it wasn't something others wanted/needed. That said, feel free to promote it however you wish. I wrote it to serve a need and saw no reason not to share it to others.

**Comment by DocDrydenn** [comment]
Oh, and sorry about jumping the gun on the merge... I forgot we were waiting on the Readme updates.

**Pull request merged**